docs/proposals/19-migrating-to-prometheus-operator.md.md

```diff
@@ -0,0 +1,530 @@
+              role: cloud-controller-manager # discover any service with this label
+        ```
+
+      ```yaml
+      apiVersion: monitoring.coreos.com/v1
+      kind: PrometheusRule
+      metadata:
+        labels:
+          cluster: shoot--project--name
+        name: cloud-controller-manager-rules
+        namespace: shoot--project--name
+      spec:
+        groups:
+        - name: cloud-controller-manager.rules
+          rules:
+          - alert: CloudControllerManagerDown
+            expr: absent(up{job="cloud-controller-manager"} == 1)
+            for: 15m
+            labels:
+              service: cloud-controller-manager
+              severity: critical
+              type: seed
+              visibility: all
+            annotations:
+              description: All infrastruture specific operations cannot be completed (e.g. creating loadbalancers or persistent volumes).
+              summary: Cloud controller manager is down.
+      ```
+
+    - Adding meta monitoring for the extensions: If the extensions need to be
+      scraped for monitoring, the extensions must bring their own [Custom
+      Resources][prom-crds].
+        - Currently the contract between extensions and gardener is that
+          anything that needs to be scraped must have the labels:
+          `prometheus.io/scrape=true` and `prometheus.io/port=<port>`. This is
+          defined [here][prom-config]. This is something that we can still
+          support with a `PodMonitor` that will scrape any pod in a specified
+          namespace with these labels.
+
+5. Add a [sidecar][grafana-sidecar] to grafana that will pickup dashboards and
+   provision them. Each dashboard gets its own configmap.
+
+    - Grafana in the control plane
+
+      - Most dashboards provisioned by Grafana are the same for each shoot
+        cluster. To avoid unnecessary duplication of configmaps, the dashboards
+        could be added once in a single namespace. These "common" dashboards can
+        then be discovered by each Grafana and provisioned.
+
+      - In some cases, dashboards are more "specific" because they are related to
+        a certain Kubernetes version.
+
+      - Contract between dashboards in configmaps and the Grafana sidecar.
+
+        - Each common dashboard will be deployed in the `monitoring` namespace as
+          a configmap. The configmap will be labeled with
+          `monitoring.gardener.cloud/dashboard-tart=shoot`.
+
+        - Each specific dashboard will be deployed in the shoot namespace. The
+          configmap will also be labeled with `monitoring.gardener.cloud/dashboard-target=shoot`.
+
+        - The grafana [sidecar][grafana-sidecar] must be [configured][sidecar-configuration] with:
+
+        ```yaml
+          env:
+          - name: METHOD
+            value: WATCH
+          - name: LABEL
+            value: monitoring.gardener.cloud/dashboard-target
+          - name: LABEL_VALUE
+            value: shoot
```

Comment:
Question: do you plan on keeping the two grafana instances (operators/users)?
If so, can you specify how dashboards can be targeted for both or either instance?
If not, can you add the removal of the second grafana instance as a goal / step to the GEP?

Reply:
I would prefer to move to 1 grafana instance. I know the logging components also use grafana. @vlvasilev @Kristian-ZH do you have any preferences?

Reply:
Ok sounds good.
For the sake of moving this GEP forward and staying flexible, I propose to use the following labels:
```
monitoring.gardener.cloud/dashboard-{seed,shoot,shoot-users}=true
```
If we want to use a single dashboard in multiple grafana instances we can add one label per target.
(I think the controller-runtime dashboard is a good example for a dashboard which we want to share between seed grafana and all shoot grafanas).

This would also allow us to keep both shoot grafana instances if we decide to do so :)

Reply:
Ideally, we would do the same for specifying the target of `*Monitor` objects and use similar labels there as well.

Reply:
I went with `monitoring.gardener.cloud/dashboard-shoot=<owner|operator>` for the shoot dashboards and `monitoring.gardener.cloud/dashboard-seed` for the seed. Motivation for this is that you cannot set multiple `LABEL`s or `LABEL_VALUE`s in the sidecar, so we must take advantage of the value of the label (shoot case).

Reply:
I will gladly remove end-user Grafana if there is no sensitive data in the logs which we don't display to the users.
Unfortunately, I don't know which component must be hidden or not.